### DISCUSSION TITLE
Unable to find routing_enums.pb.h needed to run VRPTW example

### DISCUSSION THREAD

**Discussion by onecable5781** [google/or-tools#4171] [Routing (and legacy CP) questions]
Hello,

In trying to run the example provided [here](https://developers.google.com/optimization/routing/vrptw), it seems the code needs 

`routing_enums.pb.h` but this is not available in the folder

https://github.com/google/or-tools/tree/stable/ortools/constraint_solver

So, the code does not even compile. Can you please let me know how to fix this?

Thanks.

**Comment by lperron** [comment]
It is generated from the proto file.

**Reply by onecable5781** [reply]
I see. Thank you. Is there any example of cmake for this assuming CMakeLists.txt processing will produce this file?

The example provided https://developers.google.com/optimization/introduction/cpp seems to use raw make which the main github page says is deprecated.

**Comment by lperron** [comment]
Make works if you use the binary archive.
This archive will have all generated files.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le mar. 9 avr. 2024 à 08:01, onecable5781 ***@***.***> a
écrit :

**Comment by Mizux** [comment]
you can use this template project https://github.com/or-tools/cmake_or-tools
it will rebuild or-tools and its dependencies locally...

**Comment by onecable5781** [comment]
Thank you @Mizux  and @lperron . I am afraid I am still struggling and confused.

Here is what I want to do: I want to run VRPTW on the well-known Solomon benchmark instances https://www.sintef.no/projectweb/top/vrptw/100-customers/ where the objective function is my custom distance function (which is independent of the Euclidean distance between points) and obtain the solution routes that satisfy the capacity/time window restriction at nodes and store them in my own data structure all via the same C++ program. That is, I do not want to solve the standard VRPTW where the objective function is the Eucliden distance between points. I have my own custom objective function between pair of nodes.

What I have done so far is: following the documentation https://developers.google.com/optimization/install/cpp/source_windows, I have installed OR tools by source (and not binary as @lperron seems to suggest). All installation went fine and tests passed.

Now to stuff that I am struggling with: As part of the above CMake build, already I notice that /build/examples/cpp has a cvrptw.vcxproj Visual Studio project file. Now, on opening this cvrptw project, I notice that there is a cvrptw.cc source file in the project. This file is completely different from the file provided in the documentation that explains basic VRPTW on the webpage: https://developers.google.com/optimization/routing/vrptw

Should I customize this file (since the objective function is going to be my own objective function)? Once I customize this, how can I consume this file in my other larger project?

Thank you.

**Reply by Mizux** [reply]
1) Salomon remember this pet project with the community: https://github.com/Mizux/VRP-OR-Tools
2) you should instead take a look at https://github.com/google/or-tools/tree/stable/ortools/constraint_solver/samples which are the canonical samples use in the documentation.
and this one especially https://github.com/google/or-tools/blob/stable/ortools/constraint_solver/samples/vrp_time_windows.cc

**Reply by onecable5781** [reply]
@Mizux 

I have cloned https://github.com/Mizux/VRP-OR-Tools and run it on RC208 25 node problem. (I am opening this here because there does not seem to be an issues/discussion option on that page)

The optimal solution for this problem is known to be 269.1. With the python code you have provided, the solution seems to be quite bad (assuming I am interpreting it correctly) of 583.22. Could you let me know if there are some particular settings that I should consider to get improved quality solutions?

Below I paste the output I get

(venv) :~/Downloads/vrptw/VRP-OR-Tools$ ./main.py 
instance: RC208
number of nodes: 26
number of vehicles: 25
depot: 0
Add TW: 1, [388.0,911.0]
Add TW: 2, [30.0,546.0]
Add TW: 3, [353.0,708.0]
Add TW: 4, [425.0,913.0]
Add TW: 5, [40.0,630.0]
Add TW: 6, [228.0,667.0]
Add TW: 7, [161.0,558.0]
Add TW: 8, [229.0,624.0]
Add TW: 9, [213.0,648.0]
Add TW: 10, [404.0,753.0]
Add TW: 11, [33.0,511.0]
Add TW: 12, [46.0,519.0]
Add TW: 13, [395.0,911.0]
Add TW: 14, [35.0,697.0]
Add TW: 15, [39.0,528.0]
Add TW: 16, [94.0,535.0]
Add TW: 17, [394.0,909.0]
Add TW: 18, [218.0,571.0]
Add TW: 19, [85.0,542.0]
Add TW: 20, [391.0,802.0]
Add TW: 21, [45.0,517.0]
Add TW: 22, [229.0,640.0]
Add TW: 23, [45.0,523.0]
Add TW: 24, [389.0,914.0]
Add TW: 25, [386.0,904.0]
Start solving...
Objective: 114639
Route for vehicle 0:
0 Load(0) Time(0.0,0.0) -> 24 Load(0) Time(389.0,389.0) -> 22 Load(10) Time(391.02,391.02) -> 20 Load(50) Time(393.04,393.04) -> 19 Load(60) Time(398.06,398.06) -> 18 Load(100) Time(403.464,403.464) -> 21 Load(120) Time(407.484,407.484) -> 23 Load(130) Time(409.504,409.504) -> 25 Load(160) Time(412.524,412.524) -> 9 Load(180) Time(451.594,451.594) -> 10 Load(200) Time(456.614,456.614) -> 11 Load(230) Time(458.634,458.634) -> 13 Load(270) Time(464.484,464.484) -> 15 Load(280) Time(470.334,470.334) -> 16 Load(300) Time(472.354,472.354) -> 17 Load(320) Time(477.374,477.374) -> 14 Load(340) Time(482.394,482.394) -> 12 Load(350) Time(485.414,485.414) -> 7 Load(370) Time(516.238,516.238) -> 8 Load(390) Time(521.258,521.258) -> 5 Load(400) Time(528.348,528.348) -> 3 Load(420) Time(530.368,530.368) -> 1 Load(430) Time(533.388,533.388) -> 4 Load(450) Time(540.478,540.478) -> 2 Load(490) Time(545.882,545.882) -> 6 Load(520) Time(549.902,549.902) -> 0 Load(540) Time(583.222,583.222)
Time of the route: 583.222min

Route for vehicle 1:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 2:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 3:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 4:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 5:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 6:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 7:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 8:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 9:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 10:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 11:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 12:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 13:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 14:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 15:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 16:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 17:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 18:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 19:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 20:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 21:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 22:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 23:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Route for vehicle 24:
0 Load(0) Time(0.0,0.0) -> 0 Load(0) Time(0.0,0.0)
Time of the route: 0.0min

Total time of all routes: 583.222min

**Reply by Mizux** [reply]
Do you have the solution and the 25 nodes file used ?

I just find this 100 nodes problem:
https://people.idsia.ch/~luca/macs-vrptw/problems/rc208.txt
is it this one but only use the first 25 customers ?

Also since you have TW: 4, [425.0,913.0] I don't understand how you can have less than 425 min time (here vehicle start at time `0` at location `[0,0]`)

**Reply by onecable5781** [reply]
@Mizux 

Below, at the extreme bottom of this message, I append the optimal solution which obtains 269.1. Please note that Solomon instances' objective function (and standard in VRPTW literature) only include pure transit times and do not include waiting time before start of service or service time. Perhaps that explains the difference? If so, could I request you to update the main.py file so that it solves the correct version of VRPTW?

Now, I coded up my own C++ code using your suggestions available here:https://godbolt.org/z/7bfnYPY9q

which correctly includes only the pure transit cost. Using this C++ file, I obtain a solution (can be seen in generated output.txt file) of 299

2
0 11 (t: 33.50, c: 33.50)15 (t: 49.50, c: 39.50)16 (t: 61.50, c: 41.50)13 (t: 111.00, c: 48.50)9 (t: 410.00, c: 53.50)25 (t: 459.00, c: 92.50)23 (t: 472.00, c: 95.50)21 (t: 484.00, c: 97.50)18 (t: 498.00, c: 101.50)19 (t: 513.30, c: 106.80)20 (t: 528.30, c: 111.80)22 (t: 540.30, c: 113.80)24 (t: 552.30, c: 115.80)0 (t: 597.30, c: 150.80)
0 2 (t: 30.80, c: 30.80)8 (t: 49.40, c: 39.40)5 (t: 246.00, c: 46.40)3 (t: 258.00, c: 48.40)1 (t: 366.00, c: 51.40)4 (t: 405.00, c: 58.40)6 (t: 440.30, c: 63.70)7 (t: 453.30, c: 66.70)12 (t: 494.10, c: 97.50)14 (t: 507.10, c: 100.50)17 (t: 522.10, c: 105.50)10 (t: 543.20, c: 116.60)0 (t: 584.80, c: 148.20)
Total cost: 299.00

To run the C++ file, one needs the attached input.txt file, which is read in by the C++ file's built executable:

[input.txt](https://github.com/google/or-tools/files/14948506/input.txt)

----Below is the optimal solution for RC208, 25 customer problem-----

```
Route number 1
12 14 17 16 15 13 9 11 10 2 6 7 8 4 5 3 1
Cumulative demand is 360
From 0 (40,50) Cargo 0 at 0.000000, to 12 (8,45) Cargo 20 costs/time 32.300000
Arrive 12 at 32.300000 with cumulative demand of 20, Since start at 12 is 46.000000, wait till then, Service time at 12 is 10.000000. Depart at 56.000000
From 12 (8,45) Cargo 20 at 56.000000, to 14 (5,45) Cargo 10 costs/time 3.000000
Arrive 14 at 59.000000 with cumulative demand of 30, Earliest arrival was 35.300000, Service time at 14 is 10.000000. Depart at 69.000000
From 14 (5,45) Cargo 10 at 69.000000, to 17 (0,45) Cargo 20 costs/time 5.000000
Arrive 17 at 74.000000 with cumulative demand of 50, Since start at 17 is 394.000000, wait till then, Service time at 17 is 10.000000. Depart at 404.000000
From 17 (0,45) Cargo 20 at 404.000000, to 16 (0,40) Cargo 20 costs/time 5.000000
Arrive 16 at 409.000000 with cumulative demand of 70, Earliest arrival was 94.000000, Service time at 16 is 10.000000. Depart at 419.000000
From 16 (0,40) Cargo 20 at 419.000000, to 15 (2,40) Cargo 20 costs/time 2.000000
Arrive 15 at 421.000000 with cumulative demand of 90, Earliest arrival was 39.200000, Service time at 15 is 10.000000. Depart at 431.000000
From 15 (2,40) Cargo 20 at 431.000000, to 13 (5,35) Cargo 10 costs/time 5.800000
Arrive 13 at 436.800000 with cumulative demand of 100, Earliest arrival was 395.000000, Service time at 13 is 10.000000. Depart at 446.800000
From 13 (5,35) Cargo 10 at 446.800000, to 9 (10,35) Cargo 20 costs/time 5.000000
Arrive 9 at 451.800000 with cumulative demand of 120, Earliest arrival was 213.000000, Service time at 9 is 10.000000. Depart at 461.800000
From 9 (10,35) Cargo 20 at 461.800000, to 11 (8,40) Cargo 40 costs/time 5.300000
Arrive 11 at 467.100000 with cumulative demand of 160, Earliest arrival was 33.500000, Service time at 11 is 10.000000. Depart at 477.100000
From 11 (8,40) Cargo 40 at 477.100000, to 10 (10,40) Cargo 30 costs/time 2.000000
Arrive 10 at 479.100000 with cumulative demand of 190, Earliest arrival was 404.000000, Service time at 10 is 10.000000. Depart at 489.100000
From 10 (10,40) Cargo 30 at 489.100000, to 2 (22,75) Cargo 30 costs/time 37.000000
Arrive 2 at 526.100000 with cumulative demand of 220, Earliest arrival was 30.800000, Service time at 2 is 10.000000. Depart at 536.100000
From 2 (22,75) Cargo 30 at 536.100000, to 6 (18,75) Cargo 20 costs/time 4.000000
Arrive 6 at 540.100000 with cumulative demand of 240, Earliest arrival was 228.000000, Service time at 6 is 10.000000. Depart at 550.100000
From 6 (18,75) Cargo 20 at 550.100000, to 7 (15,75) Cargo 20 costs/time 3.000000
Arrive 7 at 553.100000 with cumulative demand of 260, Earliest arrival was 161.000000, Service time at 7 is 10.000000. Depart at 563.100000
From 7 (15,75) Cargo 20 at 563.100000, to 8 (15,80) Cargo 10 costs/time 5.000000
Arrive 8 at 568.100000 with cumulative demand of 270, Earliest arrival was 229.000000, Service time at 8 is 10.000000. Depart at 578.100000
From 8 (15,80) Cargo 10 at 578.100000, to 4 (20,80) Cargo 40 costs/time 5.000000
Arrive 4 at 583.100000 with cumulative demand of 310, Earliest arrival was 425.000000, Service time at 4 is 10.000000. Depart at 593.100000
From 4 (20,80) Cargo 40 at 593.100000, to 5 (20,85) Cargo 20 costs/time 5.000000
Arrive 5 at 598.100000 with cumulative demand of 330, Earliest arrival was 40.300000, Service time at 5 is 10.000000. Depart at 608.100000
From 5 (20,85) Cargo 20 at 608.100000, to 3 (22,85) Cargo 10 costs/time 2.000000
Arrive 3 at 610.100000 with cumulative demand of 340, Earliest arrival was 353.000000, Service time at 3 is 10.000000. Depart at 620.100000
From 3 (22,85) Cargo 10 at 620.100000, to 1 (25,85) Cargo 20 costs/time 3.000000
Arrive 1 at 623.100000 with cumulative demand of 360, Earliest arrival was 388.000000, Service time at 1 is 10.000000. Depart at 633.100000
From 1 (25,85) leave at 633.100000, to 0 (40,50) costs/time 38.000000
Arrive 0 at 671.100000,
--------
Route number 2
24 25 23 21 18 19 20 22
Cumulative demand is 180
From 0 (40,50) Cargo 0 at 0.000000, to 24 (38,15) Cargo 10 costs/time 35.000000
Arrive 24 at 35.000000 with cumulative demand of 10, Since start at 24 is 389.000000, wait till then, Service time at 24 is 10.000000. Depart at 399.000000
From 24 (38,15) Cargo 10 at 399.000000, to 25 (35,5) Cargo 20 costs/time 10.400000
Arrive 25 at 409.400000 with cumulative demand of 30, Earliest arrival was 386.000000, Service time at 25 is 10.000000. Depart at 419.400000
From 25 (35,5) Cargo 20 at 419.400000, to 23 (38,5) Cargo 30 costs/time 3.000000
Arrive 23 at 422.400000 with cumulative demand of 60, Earliest arrival was 45.000000, Service time at 23 is 10.000000. Depart at 432.400000
From 23 (38,5) Cargo 30 at 432.400000, to 21 (40,5) Cargo 10 costs/time 2.000000
Arrive 21 at 434.400000 with cumulative demand of 70, Earliest arrival was 45.000000, Service time at 21 is 10.000000. Depart at 444.400000
From 21 (40,5) Cargo 10 at 444.400000, to 18 (44,5) Cargo 20 costs/time 4.000000
Arrive 18 at 448.400000 with cumulative demand of 90, Earliest arrival was 218.000000, Service time at 18 is 10.000000. Depart at 458.400000
From 18 (44,5) Cargo 20 at 458.400000, to 19 (42,10) Cargo 40 costs/time 5.300000
Arrive 19 at 463.700000 with cumulative demand of 130, Earliest arrival was 85.000000, Service time at 19 is 10.000000. Depart at 473.700000
From 19 (42,10) Cargo 40 at 473.700000, to 20 (42,15) Cargo 10 costs/time 5.000000
Arrive 20 at 478.700000 with cumulative demand of 140, Earliest arrival was 391.000000, Service time at 20 is 10.000000. Depart at 488.700000
From 20 (42,15) Cargo 10 at 488.700000, to 22 (40,15) Cargo 40 costs/time 2.000000
Arrive 22 at 490.700000 with cumulative demand of 180, Earliest arrival was 229.000000, Service time at 22 is 10.000000. Depart at 500.700000
From 22 (40,15) leave at 500.700000, to 0 (40,50) costs/time 35.000000
Arrive 0 at 535.700000,
```
